services/issue/label.go

```diff
@@ -12,49 +14,49 @@ import (
 )
 
 // ClearLabels clears all of an issue's labels
-func ClearLabels(issue *issues_model.Issue, doer *user_model.User) error {
+func ClearLabels(ctx context.Context, issue *issues_model.Issue, doer *user_model.User) error {
 	if err := issues_model.ClearIssueLabels(issue, doer); err != nil {
 		return err
 	}
 
-	notify_service.IssueClearLabels(db.DefaultContext, doer, issue)
+	notify_service.IssueClearLabels(ctx, doer, issue)
 
 	return nil
 }
 
 // AddLabel adds a new label to the issue.
-func AddLabel(issue *issues_model.Issue, doer *user_model.User, label *issues_model.Label) error {
+func AddLabel(ctx context.Context, issue *issues_model.Issue, doer *user_model.User, label *issues_model.Label) error {
 	if err := issues_model.NewIssueLabel(issue, label, doer); err != nil {
 		return err
 	}
 
-	notify_service.IssueChangeLabels(db.DefaultContext, doer, issue, []*issues_model.Label{label}, nil)
+	notify_service.IssueChangeLabels(ctx, doer, issue, []*issues_model.Label{label}, nil)
 	return nil
 }
 
 // AddLabels adds a list of new labels to the issue.
-func AddLabels(issue *issues_model.Issue, doer *user_model.User, labels []*issues_model.Label) error {
+func AddLabels(ctx context.Context, issue *issues_model.Issue, doer *user_model.User, labels []*issues_model.Label) error {
 	if err := issues_model.NewIssueLabels(issue, labels, doer); err != nil {
 		return err
 	}
 
-	notify_service.IssueChangeLabels(db.DefaultContext, doer, issue, labels, nil)
+	notify_service.IssueChangeLabels(ctx, doer, issue, labels, nil)
 	return nil
 }
 
 // RemoveLabel removes a label from issue by given ID.
-func RemoveLabel(issue *issues_model.Issue, doer *user_model.User, label *issues_model.Label) error {
-	ctx, committer, err := db.TxContext(db.DefaultContext)
+func RemoveLabel(ctx context.Context, issue *issues_model.Issue, doer *user_model.User, label *issues_model.Label) error {
+	dbCtx, committer, err := db.TxContext(ctx)
```

Comment:
I summon @wxiaoguang

Reply:
The "notification" framework has some legacy design problems.

I guess maybe the code could be like this (and the same for ReplaceLabels or others) to avoid the `ctx` trick.

```
    err := db.WithTx(ctx, func (...) {
        ... 
    })
    if err == nil {
         notify_service.......
    }
    return err
```